riscv/src/lib.rs

```diff
@@ -179,187 +119,117 @@ macro_rules! as_ref [
-) -> BTreeMap<String, String> {
-    compile_rust_crate_to_riscv(input_dir, output_dir, features).load_asm_files()
+struct CargoMetadata {
+    bins: Vec<String>,
+    use_std: bool,
 }
 
-pub fn compile_rust_crate_to_riscv_bin(
-    input_dir: &str,
-    output_dir: &Path,
-    features: Option<Vec<String>>,
-) -> PathBuf {
-    compile_rust_crate_to_riscv(input_dir, output_dir, features)
-        .executable
-        .unwrap()
-}
+impl CargoMetadata {
+    fn from_input_dir(input_dir: &str) -> Self {
+        // Calls `cargo metadata --format-version 1 --no-deps --manifest-path <input_dir>` to determine
+        // if the `std` feature is enabled in the dependency crate `powdr-riscv-runtime`.
+        let metadata = Command::new("cargo")
+            .args(as_ref![
+                OsStr;
+                "metadata",
+                "--format-version",
+                "1",
+                "--no-deps",
+                "--manifest-path",
+                input_dir,
+            ])
+            .output()
+            .unwrap();
 
-fn is_std_enabled_in_runtime(input_dir: &str) -> bool {
-    // Calls `cargo metadata --format-version 1 --no-deps --manifest-path <input_dir>` to determine
-    // if the `std` feature is enabled in the dependency crate `powdr-riscv-runtime`.
-    let metadata = Command::new("cargo")
-        .args(as_ref![
-            OsStr;
-            "metadata",
-            "--format-version",
-            "1",
-            "--no-deps",
-            "--manifest-path",
-            input_dir,
-        ])
-        .output()
-        .unwrap();
-
-    let metadata: serde_json::Value = serde_json::from_slice(&metadata.stdout).unwrap();
-    let packages = metadata["packages"].as_array().unwrap();
-    packages.iter().any(|package| {
-        package["dependencies"]
-            .as_array()
-            .unwrap()
+        let metadata: serde_json::Value = serde_json::from_slice(&metadata.stdout).unwrap();
+        let packages = metadata["packages"].as_array().unwrap();
+
+        // Is the `std` feature enabled in the `powdr-riscv-runtime` crate?
+        let use_std = packages.iter().any(|package| {
+            package["dependencies"]
+                .as_array()
+                .unwrap()
+                .iter()
+                .any(|dependency| {
+                    dependency["name"] == "powdr-riscv-runtime"
+                        && dependency["features"]
+                            .as_array()
+                            .unwrap()
+                            .contains(&"std".into())
+                })
+        });
+
+        let bins = packages
             .iter()
-            .any(|dependency| {
-                dependency["name"] == "powdr-riscv-runtime"
-                    && dependency["features"]
-                        .as_array()
-                        .unwrap()
-                        .contains(&"std".into())
+            .flat_map(|package| {
+                package["targets"]
```

Comment:
where was this before? I don't see it in the removed parts

Reply:
This is new code.

The previous version used the JSON from `cargo -Zunstable-options --build-plan` to get details from every compilation unit and find the assembly files. But it is an unstable feature with a heavy run cost (it needs a new target directory to run, and redo a lot of work).

Now that we just need to find the executables, so we use the stable and and lightweight JSON from `cargo metadata` (that was already been called to figure out if this is a `std` or `no-std` build).